lib/compress/zstd_compress.c

```diff
@@ -5645,7 +5644,15 @@ size_t ZSTD_compressStream2( ZSTD_CCtx* cctx,
 
     /* transparent initialization stage */
     if (cctx->streamStage == zcss_init) {
-        FORWARD_IF_ERROR(ZSTD_CCtx_init_compressStream2(cctx, endOp, input->size), "CompressStream2 initialization failed");
+        size_t const inputSize = input->size - input->pos;  /* no obligation to start from pos==0 */
+        if ( (cctx->requestedParams.inBufferMode == ZSTD_bm_stable) /* input is presumed stable, across invocations */
+          && (endOp == ZSTD_e_continue)                             /* no flush requested, more input to come */
+          && (inputSize < ZSTD_BLOCKSIZE_MAX) ) {                   /* not even reached one block yet */
+            /* just wait, allows lazy adaptation of compression parameters */
+            cctx->expectedInBuffer = *input;
+            return ZSTD_FRAMEHEADERSIZE_MIN(cctx->requestedParams.format);  /* at least some header to produce */
+        }
```

Comment:
I don't think that this is correct. I get that the use case of streaming + stable input is somewhat contrived, and that it isn't super useful in practice. But...

Zstd guarantees to make forward progress. Either consume some input or produce some output. This breaks that contract, since `input` is not modified.

A loop constructed like this should not infinite loop:

```
while (!(input empty || output empty))
  ZSTD_compressStream(cctx, &output, &input);
```

Instead, we should save the initial `input->pos`, and update `input->pos = input->size`.

Additionally, we shouldn't directly set `cctx->expectedInBuffer`, `ZSTD_setBufferExpectations(cctx, output, input)` should be used instead. The call on line 5656 can probably be moved up to handle both lazy & non-lazy adaptation, but only for the first call of lazy adaptation.

Lastly, since we are now modifying the input, we should make sure that we are calling `FORWARD_IF_ERROR(ZSTD_checkBufferStability(cctx, output, input, endOp), "invalid buffers");` in the lazy-adaptation mode.

Reply:
> Zstd guarantees to make forward progress. 

Forward progress was indeed one of my concerns when writing that code.

Some nuances here : there are limits to forward progress "guarantees" that `zstd` provides, even with current code.
As a perhaps obvious example, if a user continuously presents an empty input, then there is nothing to do. Same thing when a user continuously presents an output buffer with no more available space to flush into, then the streaming interface cannot progress. In both cases, repetitively calling `ZSTD_compressStream()` will just loop forever.

But it's okay, I guess we can classify above scenarios as user errors, that users should be able to debug, while the scenario you advocate for is different. What you mean is that, if it is possible to make some forward progresses, aka there is room in the output buffer and there is some input to ingest, then `zstd` make some forward progress.
In this instance, instead of returning immediately without doing anything, `zstd` should fake progress by pretending having consumed the input, even though it hasn't, so that the loop condition you suggest, which relies on forward progress, can exit.

I think that's fair, and that's what latest commit in this PR provides.

Reply:
> we shouldn't directly set `cctx->expectedInBuffer`, `ZSTD_setBufferExpectations(cctx, output, input)` should be used instead.

`ZSTD_setBufferExpectations()` relies on `cctx->appliedParams` which is _not initialized yet_ since at this stage `ZSTD_CCtx_init_compressStream2()` has not been invoked yet.